packages/venia-concept/src/components/ProductImageCarousel/carousel.js

```diff
@@ -1,136 +1,108 @@
-            currentImage: PropTypes.string,
-            imageContainer: PropTypes.string,
-            'chevron-left': PropTypes.string,
-            'chevron-right': PropTypes.string
-        }),
-        images: PropTypes.arrayOf(
-            PropTypes.shape({
-                label: PropTypes.string,
-                position: PropTypes.number,
-                disabled: PropTypes.bool,
-                file: PropTypes.string.isRequired
-            })
-        ).isRequired
-    };
+const DEFAULT_IMAGE_WIDTH = 640;
 
-    state = {
-        activeItemIndex: 0,
-        currentImageLoaded: false
-    };
+const Carousel = props => {
+    const classes = mergeClasses(defaultClasses, props.classes);
 
-    updateActiveItemIndex = index => {
-        this.setState({ activeItemIndex: index });
-    };
+    const [carouselState, carouselApi] = useCarousel(props.images);
+    const { activeItemIndex, currentImage, sortedImages } = carouselState;
+    const { handlePrevious, handleNext, setActiveItemIndex } = carouselApi;
 
-    // The spec does not guarantee a position parameter,
-    // so the rule will be to order items without position last.
-    // See https://github.com/magento/graphql-ce/issues/113.
-    // Memoize this expensive operation based on reference equality
-    // of the `images` array. Apollo cache should return a new array
-    // only when it does a new fetch.
-    sortAndFilterImages = memoize(items =>
-        items
-            .filter(i => !i.disabled)
-            .sort((a, b) => {
-                const aPos = isNaN(a.position) ? 9999 : a.position;
-                const bPos = isNaN(b.position) ? 9999 : b.position;
-                return aPos - bPos;
-            })
-    );
-
-    get sortedImages() {
-        const { images } = this.props;
-        return this.sortAndFilterImages(images);
-    }
+    const handleLeftChevron = useCallback(() => {
+        handlePrevious();
+    }, [handlePrevious]);
 
-    leftChevronHandler = () => {
-        const sortedImages = this.sortedImages;
-        const { activeItemIndex } = this.state;
-        activeItemIndex > 0
-            ? this.updateActiveItemIndex(activeItemIndex - 1)
-            : this.updateActiveItemIndex(sortedImages.length - 1);
-    };
+    const handleRightChevron = useCallback(() => {
+        handleNext();
+    }, [handleNext]);
 
-    rightChevronHandler = () => {
-        const sortedImages = this.sortedImages;
-        const { activeItemIndex } = this.state;
-        this.updateActiveItemIndex((activeItemIndex + 1) % sortedImages.length);
-    };
+    const handleThumbnailClick = useCallback(index => {
+        setActiveItemIndex(index);
+    }, []);
 
-    getChevron = direction => (
-        <button
-            onClick={this[`${direction}ChevronHandler`]}
-            className={this.props.classes[`chevron-${direction}`]}
-        >
-            <Icon src={ChevronIcons[direction]} size={40} />
-        </button>
+    const getChevron = useCallback(
```

Comment:
This is a bit too clever. I think I may just `useMemo` for both chevrons and call it a day.